### PR TITLE
refactor DdgGenerator.scala

### PR DESCRIPTION
**Pull request by Liyw979**
**Slightly modified one test case of kotlin dataflow**
I am not sure the refator is 100% correct, needs further discussion.

### CODE REVIEW

File: dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala [hunk 1]
```diff
@@ -31,96 +29,77 @@ class DdgGenerator(semantics: Semantics) {
...
-          .filterNot(x => numberToNode.get(x).contains(use))
-          .flatMap(numberToNode.get)
-          .foreach { inElemNode =>
-            addEdge(inElemNode, ret, nodeToEdgeLabel(inElemNode))
-          }
-        if (inElements.isEmpty) {
-          addEdge(method, ret)
-        }
-      }
+      val uses = ret.astChildren.collect { case x: Expression => x }.toList
```
Comment (mpollmeier) [inline dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala:100]:
```suggestion
      val uses = ret.astChildren.collectAll[Expression].toList
```

**Review by mpollmeier — COMMENTED** [review]
Thank you! From a coding style perspective this is a step forward IMO. 
@fabsx00 can you check if the test outcome change is ok?

**Comment by Liyw979** [comment]
According to the comment
https://github.com/joernio/joern/blob/43d332f6db7563b1f5dd3824cf032c841913f264/dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala#L45
A node can only be linked by the ```method``` node if this node has no other reachingdef edge
However, the old ddg generator does not do the same
image

File: dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala (hunk 1 above)
Comment (Liyw979) [reply dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala:100]: Looks like a lot of tests are broken. I will try fix them

**Comment by Liyw979** [comment]
@fabsx00 In the first commit, I used 
```
    def incomingDefsForNode(node: StoredNode): Set[StoredNode] = {
      in(node).toSet
```
which doesn't follows the original logic
In the last commit, It's changed to use the origin logic
```
    def incomingDefsForNode(current: StoredNode, use: StoredNode): Set[StoredNode] = {
      in(current).toSet
```

**Comment by Liyw979** [comment]
This pr is still needs improvement. leave this pr open.

**Comment by Liyw979** [comment]
Closing this since bugs I found in DdgGenerator have being fixed.

**Pull request closed without merging**